README.md

```diff
@@ -1,28 +1,204 @@
+
+`url-mapper` is helpful when creating router packages for frameworks or can be used as router itself.
+It allows you to outsource work with url (mapping, parsing, stringifying) and concentrate on wiring up things related to your favorite framework.
+
+### Example
 
-const someFunc = function (data) {
-  data // {url: '/foo/123?foo=bar', path: '/foo', params: {id: '123'}, query: {foo: 'bar'}}
-};
+```js
+import React from 'react';
+import ReactDOM from 'react-dom';
+import Mapper from 'url-mapper';
+import { CoreApp, ComponentA, ComponentB, Component404 } from './components';
 
-const notFoundFunc = function (data) { };
+const urlMapper = Mapper();
 
-route(location.origin + '/foo/123', {
-  '/foo/:id': someFunc,
-  '*': notFoundFunc
+var matchedRoute = urlMapper('/bar/baz/:42', { // routable part of url
+  '/foo/:id': ComponentA,
+  '/bar/:list/:itemId': ComponentB,
+  '*': Component404
 });
+
+if (matchedRoute) {
+  const Component = matchedRoute.match; // ComponentB
+  const props = matchedRoute.values; // { list: 'baz', itemId: 42 }
+
+  ReactDOM.render(
+    <CoreApp>
+      <Component {...props} />
+    </CoreApp>
+  );
+}
 ```
-**url-mapper** passes an object representing the parsed URL to your route callbacks. Given the URL `http://www.bigapp.com/foo/123?bar=baz` and a matching route `/foo/:id`, your callbacks will receive:
+
+See [`cerebral-router`](https://github.com/christianalfoni/cerebral-router) as example of building own router solution on top of `url-mapper`.
+
+## API
+
+### Main module
+
+At top level `url-mapper` module exports factory which returns default implementation of `URL` <==> `route(params)` converter.
+
+#### Factory
+##### Usage
+
 ```js
-{
-  url: '/foo/123?bar=baz',
-  path: '/foo/123',
-  params: {id: 123},
-  query: {bar: 'baz'}
-}
+var urlMapper = require('url-mapper');
+var mapper = urlMapper(options);
 ```
 
-This library just allows you to map a url to a function. It allows dynamic segments and you can pass a full url or just the path. It's designed to be used with other libraries to manage URLs in your applications.
+##### Arguments
+
+<table>
+  <tr>
+    <th>Param</th><th>Type</th><th>Details</th>
+  </tr>
+  <tr>
+    <td>options</th><td><code>Object</code></td>
+    <td>Options passed to converter.
+      <table>
+        <tr>
+          <th>Property</th><th>Type</th><th>Details</th>
+        </tr>
+        <tr>
+          <td>query</td><td><code>Boolean</code></td>
+          <td>Enables converting values not defined in route as query in URL Object Notation</td>
+        </tr>
+      </table>
+    </td>
+  </tr>
+</table>
+
+##### Returns
+
+`Object` - Object with `parse`, `stringify` and `map` methods.
+
+Returned methods deals with Express-style route definitions and cleaned routable part of url (without origin, base path, leading hash symbol).
+
+Params defined in route is mapped to the same named properties in `values` Object with help of `path-to-regexp` module.
```

Comment:
Params defined in route are mapped to the same named properties in the `values` Object with help of `path-to-regexp` module.